packages/api/cms-api/src/dam/files/dam-items.service.ts

```diff
@@ -1,47 +1,49 @@
 import { Injectable } from "@nestjs/common";
 
-import { DamItem } from "./dam-items.resolver";
-import { DamItemsArgs } from "./dto/dam-items.args";
+import { DamScopeInterface } from "../types";
+import { DamItemInterface } from "./dam-items.resolver";
+import { DamItemsArgsInterface } from "./dto/dam-items.args";
 import { FilesService } from "./files.service";
 import { FoldersService } from "./folders.service";
 
 @Injectable()
 export class DamItemsService {
     constructor(private readonly foldersService: FoldersService, private readonly filesService: FilesService) {}
 
-    async findAndCount({
-        folderId,
-        includeArchived,
-        filter,
-        sortColumnName,
-        sortDirection,
-        offset,
-        limit,
-    }: DamItemsArgs): Promise<[typeof DamItem[], number]> {
-        const [folders, foldersTotalCount] = await this.foldersService.findAndCount({
-            parentId: folderId,
-            includeArchived,
-            filter: { searchText: filter?.searchText },
-            sortDirection,
-            sortColumnName,
-            offset,
-            limit,
-        });
+    async findAndCount(
+        { folderId, includeArchived, filter, sortColumnName, sortDirection, offset, limit }: Omit<DamItemsArgsInterface, "scope">,
+        scope?: DamScopeInterface,
```

Comment:
I deliberately added scope as a second argument (instead of in the first argument object) so that using `nonEmptyScopeOrNothing(scope)` (see resolvers) is more prominent. We may change it if you prefer so.

Reply:
It is questionable if a service should use the resolver dtos at all.

Imho:

- either a fully independent service containing business logic (merging two findAndCount as yoiu do here might not count as business logic)
- or not service at all

But all in all, it's a private service (right?) so it doesn't really matter much.

Reply:
I enjoy such discussions :)

let's take your create redirect exampe, here is the current resolver (that doesn't use a service):
```
        @Mutation(() => Redirect)
        async createRedirect(
            @Args("scope", { type: () => Scope, defaultValue: hasNonEmptyScope ? undefined : {} }) scope: typeof Scope,
            @Args("input", { type: () => RedirectInput }) input: RedirectInputInterface,
        ): Promise<RedirectInterface> {
            if (!(await this.redirectService.isRedirectSourceAvailable(input.source, nonEmptyScopeOrNothing(scope)))) {
                throw new CometValidationException("Validation failed");
            }

            const tranformedInput = plainToInstance(RedirectInput, input);

            const errors = await validate(tranformedInput, { whitelist: true, forbidNonWhitelisted: true });

            if (errors.length > 0) {
                throw new CometValidationException("Validation failed", errors);
            }

            const entity = this.repository.create({
                scope: nonEmptyScopeOrNothing(scope),
                ...tranformedInput,
                target: tranformedInput.target.transformToBlockData(),
            });
            await this.repository.persistAndFlush(entity);
            return this.repository.findOneOrFail(entity.id);
        }
```

turning that into the following I would not recommend, because:
- this manual plainToInstance + validation is highly resolver specific (any only needed here because of dynamic scope)
- persistAndFlush is not always wanted (eg. to do it only once after a loop)
- the return value (via findOneOrFail) is not always wanted (causes additional queries even if not used at all)
```
@Mutation(() => Redirect)
        async createRedirect(
            @Args("scope", { type: () => Scope, defaultValue: hasNonEmptyScope ? undefined : {} }) scope: typeof Scope,
            @Args("input", { type: () => RedirectInput }) input: RedirectInputInterface,
        ): Promise<RedirectInterface> {
            return this.redirectService.createRedirect(scopoe, input);
        }
```
what is left is isRedirectSourceAvailable (already in service, contains business logic) and `repository.create`. Having create as alias for repository.create in the service is not useful at all. The only thing that *could* make sense is a `redirectService.validateRedirectSourceAvailableAndCreate` (or a bit shorter name). But I don't think validation and creation should be in a single method as they are two different steps.

So, for your CLI command example: `repository.create()` seems a pretty good fit, imho.

Reply:
What I found particularly annoying about the redirect example was that the redirect contains block JSON. The code looked like this:

```
const redirect = this.repository.create({
    id: uuid(),
    sourceType: RedirectSourceTypeValues.path,
    source: row["source"],
    target: linkBlock
        .blockInputFactory({
            attachedBlocks: [
                {
                    type: "internal",
                    props: {
                        targetPageId: node.id,
                    },
                },
            ],
            activeType: "internal",
        })
        .transformToBlockData(),
    active: true,
    generationType: RedirectGenerationType.manual,
    createdAt: new Date(),
    updatedAt: new Date(),
    comment: comment,
});
```

I would have found it much simpler to have a service like this:

```
const redirect = this.redirectsService.create({
    sourceType: RedirectSourceTypeValues.path,
    source: row["source"],
    targetType: "internal",
    internalTargetProps: {
        targetPageId: node.id,
    },
    generationType: RedirectGenerationType.manual,
    comment: comment,
});
```

A service like this is easy to understand and you don't have to know anything about the block system. But that's of course a very specific example.

Generally, I just feel that not implementing services may lead to more code duplication and may make the development of applications more effortful. But of course, you have a lot more experience than me so I may be wrong

Reply:
Yes, that manual block creation is not what you want do do from the outside. The service.create (you wished was there) would have have done that for you though, as in the use case for the resolver not only internal links must be supported.

What you need is a new service method like `createInternalRedirect` that even accepts directly `targetPageId` or so. This makes absolutely sense and should be done!

The really, really hard thing here is that this is in a library and you should know what potentially could be needed by the applications.

(you can leave out the id/createdAt/updatedAt as those have default values)

> Generally, I just feel that not implementing services may lead to more code duplication and may make the development of applications more effortful

I never said services are a bad thing or should not be used and code should be duplicated instead. I'm only skeptical for services that provide no value, typical examples:
- "repository wrapper service": providing incomplete access the repository, eg. `findOne(id: string) { this.repository.findOne(id); }` without any additional value -  instead of using it directly
- "resolver methods service": moving all resolver methods to a service with one-liners left in the resolver

Creating such repository wrapper services or resolver methods services are actually code duplication.

> I may be wrong

I may be wrong too :)

Reply:
I believe that one issue with using the Redirect repository in the application code was the absence of a Redirect entity, as it is created dynamically. One would have to know that the repository can be injected using a string `@InjectRepository("Redirect")`. We may add an example in demo and/or starter.

> What you need is a new service method like createInternalRedirect that even accepts directly targetPageId or so. This makes absolutely sense and should be done!

This is a good idea!